tests/test_gpu_common.py

```diff
@@ -74,6 +75,89 @@ def test_autopad_with_platform_switch(self):
 
 
+class TestEnvironmentVariables:
+    """
+    Test that environment variables are correctly handled.
+    """
+
+    @pytest.mark.parametrize('env_variables', [{"cuda_visible_devices": "1"},
+                                               {"cuda_visible_devices": "1,2"},
+                                               {"cuda_visible_devices": "1,0"},
+                                               {"rocr_visible_devices": "1"},
+                                               {"hip_visible_devices": " 1"}])
+    def test_visible_devices(self, env_variables):
+        """
+        Test that physical device IDs used for querying memory on a device via
+        nvidia-smi correctly account for visible-device environment variables.
+        """
+        grid = Grid(shape=(10, 10))
+        u = Function(name='u', grid=grid)
+
+        eq = Eq(u, u+1)
+
+        with switchenv(**env_variables):
+            op1 = Operator(eq)
+
+            argmap1 = op1.arguments()
+            # All variants in parameterisation should yield deviceid 1
+            assert argmap1._physical_deviceid == 1
+
+        # Make sure the switchenv doesn't somehow persist
+        for i in ("CUDA", "ROCR", "HIP"):
+            assert f"{i}_VISIBLE_DEVICES" not in os.environ
+
+        # Check that physical deviceid is 0 when no environment variables set 
+        op2 = Operator(eq)
+
+        argmap2 = op2.arguments()
+        # Default physical deviceid expected to be 0
+        assert argmap2._physical_deviceid == 0
+
+
+    @pytest.mark.parallel(mode=2)
+    @pytest.mark.parametrize('visible_devices', ["1,2", "1,0", "0,2,3"])
+    def test_visible_devices_mpi(self, visible_devices, mode):
+        """
+        Test that physical device IDs used for querying memory on a device via
+        nvidia-smi correctly account for visible-device environment variables
+        when using MPI.
+        """
+
+        grid = Grid(shape=(10, 10))
+        rank = grid.distributor.myrank
+        u = Function(name='u', grid=grid)
+
+        eq = Eq(u, u+1)
+
+        with switchenv(cuda_visible_devices=visible_devices):
+            op1 = Operator(eq)
+            argmap1 = op1.arguments()
+
+            devices = [int(i) for i in visible_devices.split(',')]
+
+            assert argmap1._physical_deviceid == devices[rank]
+
+        # In default case, physical deviceid will equal rank
+        op2 = Operator(eq)
+        argmap2 = op2.arguments()
+        assert argmap2._physical_deviceid == rank
+
+    def test_visible_devices_with_devito_deviceid(self):
+        """Test interaction between CUDA_VISIBLE_DEVICES and DEVITO_DEVICEID"""
+        grid = Grid(shape=(10, 10))
+        u = Function(name='u', grid=grid)
+
+        eq = Eq(u, u+1)
+
+        with switchenv(cuda_visible_devices="1,3"), switchconfig(deviceid=1):
+            op = Operator(eq)
+
+            argmap = op.arguments()
+            # deviceid should see the world from within CUDA_VISIBLE_DEVICES
```

Comment:
I don't think that is the wanted behavior.  `DEVITO_DEVICEID` is the id of the device not the index withing the devices so this isn't compatible with what the configuration does and might lead to problems for users currently using deviceid